nejma/layers.py

```diff
@@ -73,12 +78,73 @@ def validate_name(self, name):
             + "only alphanumerics and underscores are accepted"
         )
 
-    def clean_expired(self):
+    async def clean_expired(self):
         for group in self.groups:
             # Can't change dict size during iteration
             for channel in list(self.groups.get(group, {})):
                 if channel.is_expired():
                     del self.groups[group][channel]
 
 
+class RedisLayer(ChannelLayer):
+    def __init__(self, redis_host="redis://localhost", prefix="group"):
+        self.initialized = False
+        self.send = None
+        self.redis_host = redis_host
+        self.prefix = prefix
+
+    ### utility functions ###
+
+    def group_prefix(self, group):
+        group = str(group)
+        return f"{self.prefix}_{group}" if not group.startswith(
+            f"{self.prefix}_") else f"{group}"
+
+    async def _initialize(self):
+        self.redis: aioredis.Redis = await aioredis.create_redis_pool(self.redis_host)
+        self.initialized = True
+
+    ### add/remove interface ###
+
+    async def add(self, group_name, channel, send=None):
+        if not self.initialized:
+            await self._initialize()
+        assert self.validate_name(group_name), "Invalid group name"
+        if isinstance(channel, Channel):
+            channel = channel.name
+        group_name = self.group_prefix(group_name)
+        await self.redis.hset(group_name, channel, 1)
+
+    async def remove_channel(self, channel):
+        if not self.initialized:
+            await self._initialize()
+        if isinstance(channel, Channel):
+            channel = channel.name
+        _, keys = await self.redis.scan(match=self.group_prefix("*"))
+        for key in keys:
+            await self.redis.hdel(key, channel)
+
+    async def remove(self, group_name, channel):
+        if not self.initialized:
+            await self._initialize()
+        group_name = self.group_prefix(group_name)
+        await self.redis.hdel(group_name, channel)
+
+    async def flush(self):
+        if not self.initialized:
+            await self._initialize()
+        _, keys = await self.redis.scan(match=self.group_prefix("*"))
+        for key in keys:
```

Comment:
same here `async for`

Reply:
I think only scan-line commands support it.
https://aioredis.readthedocs.io/en/v1.2.0/start.html#python-3-5-async-with-async-for-support

Reply:
Well, for all scan commands I will use iscan then, which is iterative scan.
But for hgetall it can't be done